### PR TITLE
Remove set_default_dtype call in jit_metaprogramming_utils.py (old)

### LINKED ISSUE #68972: default dtype is incorrectly set to torch.double when running test_ops.py
Imports into test files should not be setting the default dtype. We should add asserts to the test files after imports that the default dtype is float32. 

@eellison maybe leaking from a jit include? 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #68972

NOTE: This PR may be superseded by #69624, which avoids the explicit `dtype=torch.double` that was added to many tests in this PR.

cc @mruberry

### CODE REVIEW

File: torch/testing/_internal/jit_metaprogramming_utils.py [hunk 1]
```diff
@@ -57,9 +56,9 @@
     ('max_pool2d', (S, S, S, S), (2, 1), '', (True, 'aten::max_pool2d_with_indices')),
     ('max_pool2d', (S, S, S, S), (2, 1, 1, 1, False, True), 'with_indices', (True, 'aten::max_pool2d_with_indices')),
     ('max_pool3d', (S, S, S, S, S), (2, 1)),
-    ('max_unpool1d', torch.tensor([[[2., 4]]]), (torch.tensor([[[1, 3]]]), 2, 2, 0)),
-    ('max_unpool2d', torch.tensor([[[[2., 4]]]]), (torch.tensor([[[[1, 3]]]]), 2, 2, 0)),
-    ('max_unpool3d', torch.tensor([[[[[2., 4]]]]]), (torch.tensor([[[[[1, 3]]]]]), 2, 2, 0)),
+    ('max_unpool1d', torch.tensor([[[2., 4]]], dtype=default_dtype), (torch.tensor([[[1, 3]]]), 2, 2, 0)),
```
Comment (mruberry) [inline torch/testing/_internal/jit_metaprogramming_utils.py:59]: Whoa, I had no idea this list existed. @kurtamohler are these tests redundant with our OpInfo-based jit tests?

cc @suo @eellison 

If we can remove this list that'd be awesome. Seems like it must be mergeable with our OpInfo-based tests?

**Comment by mruberry** [comment]
@kurtamohler this is super cool -- see my question for whether we can make it even cooler.

Also would you also add an assert in test_ops.py after its imports that the default_dtype is torch.float32?

File: torch/testing/_internal/jit_metaprogramming_utils.py (hunk 1 above)
Comment (eellison) [reply torch/testing/_internal/jit_metaprogramming_utils.py:59]: In the long-term, yes, I would like to remove them. In the short term, it's tough, I think we need to separate out all of the jit different testing variations into separate tests. As is, it's pretty common for one test case to disable testing for all downstream tests. For example, Conv empty input breaks our alias analysis test, so now all jit tests are disabled for conv. Id be worried about getting rid of these tests which are pretty critical for autodiff testing because the OpInfo tests so frequently are disabled. 

Opinfos have a lot of other, much better properties but I feel more safe having these tests around in the short term

Comment (mruberry) [reply torch/testing/_internal/jit_metaprogramming_utils.py:59]: Would you file an issue to track this refactoring and your concerns/thoughts about how best to do it, then @eellison? We can apply the "better engineering" label to it.

**Comment by kurtamohler** [comment]
I've added the assert to `test_ops.py` and `test_nn.py`, and I've fixed the failing tests that I knew about. We'll see if any others fail

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -5423,10 +5423,23 @@ def sample_inputs_fmod_remainder(op_info, device, dtype, requires_grad, *, autod
...
-            ((S, S, S), 1.5, False),
-            ((), 1.5, False),
-        )
+        if dtype.is_floating_point or dtype.is_complex:
+            samples = (
+                ((S, S, S), 1.5, False),
+                ((), 1.5, False),
+            )
+        else:
+            # NOTE: Given an integer tensor and a non-tensor number, fmod and
```
Comment (kurtamohler) [inline torch/testing/_internal/common_methods_invocations.py:5432]: Just wanted to draw attention to this oddity. Maybe there's a better way to handle it than this?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -1296,8 +1296,8 @@ def make_singular_matrix_batch_base(size, rank):
 
         with torch.no_grad():
             n = size[-1]
-            a = make_arg(size[:-2] + (n, rank)) / 10
-            b = make_arg(size[:-2] + (rank, n)) / 10
+            a = make_arg(size[:-2] + (n, rank), dtype=torch.double) / 10
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1299]: This use of double is interesting because it overrides the dtype requested for the sample inputs. Is it necessary and will the resulting sample inputs be double because of it?

cc @Lezcano

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5461]: Do we still want them to return a double tensor? Like what if we just let them return a float32 tensor in these cases?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -7458,7 +7471,7 @@ def sample_inputs_pairwise_distance(op_info, device, dtype, requires_grad, **kwa
     ]
 
     return [
-        SampleInput(make(shape), args=(make(shape),), kwargs=kwargs) for shape, kwargs in shapes_and_kwargs
+        SampleInput(make(shape), args=(make(shape, dtype=torch.double),), kwargs=kwargs) for shape, kwargs in shapes_and_kwargs
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7474]: Why override the dtype here?

File: torch/testing/_internal/common_nn.py [hunk 5]
```diff
@@ -102,6 +102,7 @@ def get_weight(m):
     dict(
         module_name='Linear',
         constructor_args=(10, 8),
+        constructor_kwargs=dict(dtype=torch.double),
```
Comment (mruberry) [inline torch/testing/_internal/common_nn.py:105]: cc @jbschlosser -- you probably want these to be constructed in float not double?

**Comment by kurtamohler** [comment]
@mruberry, when I removed the `set_default_dtype()` call in `jit_metaprogramming_utils.py`, all of the tests that I've edited in this PR were failing because they expected doubles. My thought was that it would be safer to continue testing with doubles wherever possible, rather than to change them to use flotas. But maybe that was not the right decision. Would it be better to make all these tests expect floats instead?

**Comment by mruberry** [comment]
> @mruberry, when I removed the `set_default_dtype()` call in `jit_metaprogramming_utils.py`, all of the tests that I've edited in this PR were failing because they expected doubles. My thought was that it would be safer to continue testing with doubles wherever possible, rather than changing them to use floats, which is why I made all the explicit `dtype=torch.double` changes. But maybe that was not the right decision. Would it be better to make all these tests expect floats instead?

holy cow! I'm very curious about the failures. For the comments left above I think the intent is to test in float32 (since that's the most commonly used datatype). So it'd be good to better understand why double was necessary and yes, possibly to fix it.

**Comment by kurtamohler** [comment]
> So it'd be good to better understand why double was necessary and yes, possibly to fix it.

Agreed, I'll put this PR in draft mode and create an alternate PR with all the changes necessary to make float testing pass, so we can compare them easily.
